### PR TITLE
Optimize DistMuon packing copies

### PR DESCRIPTION
**Pull request by jadenmach2**
## Motivation
  DistMuon’s redistribution runtime packs and unpacks tensor regions around its all-to-all collectives. Each region was copied with an individual  `Tensor.copy_()` call.
  For parameters split into multiple regions, this produces several small copy operations and GPU launches for each redistribution phase. Profiling the Kimi  K2.7 DistMuon configuration on four AMD MI250 GPUs showed an average of 493 asynchronous copies and 1,474 non-RCCL kernel launches per rank during the optimizer step.

  ## Benchmark
  Measured on four AMD MI250 GPUs with ROCm using the Kimi K2.7 debug model with DistMuon, FSDP degree 4, and expert-parallel degree 2.
  Values are averages across four ranks for the same profiled optimizer step.
  | Metric | Baseline | Batched copies | Change |
  | --- | ---: | ---: | ---: |
  | Async copies per rank | 493 | 423 | -14.2% |
  | Non-RCCL kernel launches per rank | 1,474 | 1,421.5 | -3.6% |
  | DistMuon GPU span | 93.12 ms | 92.54 ms | -0.6% |
  | DistMuon CPU span | 92.29 ms | 92.64 ms | +0.4% |
  | RCCL launches per rank | 8 | 8 | unchanged |

  The optimization measurably reduces copy and launch overhead. End-to-end optimizer latency remains effectively unchanged within run-to-run noise.

### CODE REVIEW

File: torchtitan/distributed/flex_shard/_optimizer_reshard_runtime.py [hunk 1]
```diff
@@ -757,16 +770,48 @@ def _finalize_redistributed(
...
+    foreach_compatible = all(
+        destination.layout is torch.strided
+        and source.layout is torch.strided
+        and destination.shape == source.shape
+        and destination.device == source.device == reference_device
+        and destination.dtype == source.dtype == reference_dtype
+        for destination, source in zip(destinations, sources, strict=True)
+    )
+    if foreach_compatible:
+        torch._foreach_copy_(destinations, sources)
```
Comment (weifengpy) [inline torchtitan/distributed/flex_shard/_optimizer_reshard_runtime.py]: this is good change! we can make it default, without considering backward compatibility. i do think optimizer has uniform dtype, layout, device

how about this via a minimal lines of code, example https://github.com/pytorch/torchtitan/pull/4271 (my cheap ai slop to make things concrete)
* only do foreach_compatible logic
* `torch._foreach_copy_` check compatbility
* rename to `_batched_copy_`

Comment (jadenmach2) [reply torchtitan/distributed/flex_shard/_optimizer_reshard_runtime.py]: @weifengpy Pushed the requested simplification, following the approach demonstrated in #4271:

  • removed the compatibility checks and fallback;
  • renamed the helper to _batched_copy_;
  • updated tests for dtype conversion and shape mismatch behavior.

  I also reran the four-GPU Kimi DistMuon FSDP+EP E2E test on AMD MI250 (ROCm). It completed successfully with matching loss and gradient norm across all ranks.  Thanks for the review!

**Review by weifengpy — APPROVED** [review]
thanks for the quick follow up! approved

we have integration tests that exercise the code path on real gpus
* kimi_k2_5_muon_pp+fsdp+ep on 4 GPUs
* kimi_k2_5_muon_fsdp+ep on 8
 
`test_optimizer_reshard_runtime.py` adds value by verifying empty input and special case of single parameters. but I am fine with not having such coverage. is it possible to delete `test_optimizer_reshard_runtime.py` entirely? also unit test only runs on cpu

**Review by tianyu-l — APPROVED** [review]
